src/Framework/Framework/Binding/DotvvmCapabilityProperty.cs

```diff
@@ -336,20 +336,22 @@ internal static IControlAttributeDescriptor InitializeArgument(ICustomAttributeP
                     dotvvmProperty = new DotvvmProperty(propertyName, type, declaringType, boxedDefaultValue, false, attributeProvider);
                     dotvvmProperty.OwningCapability = declaringCapability;
 
-                    var isNullable = propertyType.IsNullable() || propertyType.UnwrapValueOrBinding().IsNullable();
-
-                    if (!defaultValue.HasValue && !isNullable)
-                        dotvvmProperty.MarkupOptions.Required = true;
-
-                    if (typeof(IBinding).IsAssignableFrom(propertyType))
-                        dotvvmProperty.MarkupOptions.AllowHardCodedValue = false;
-                    else if (!typeof(ValueOrBinding).IsAssignableFrom(propertyType.UnwrapNullableType()))
-                        dotvvmProperty.MarkupOptions.AllowBinding = false;
-
-                    if (typeof(IDotvvmObjectLike).IsAssignableFrom(type) ||
-                        typeof(ITemplate).IsAssignableFrom(type) ||
-                        typeof(IEnumerable<IDotvvmObjectLike>).IsAssignableFrom(type))
-                        dotvvmProperty.MarkupOptions.MappingMode = MappingMode.Both;
+                    if (!attributeProvider.IsDefined(typeof(MarkupOptionsAttribute), true))
+                    {
+                        var isNullable = propertyType.IsNullable() || propertyType.UnwrapValueOrBinding().IsNullable();
+                        if (!defaultValue.HasValue && !isNullable)
+                            dotvvmProperty.MarkupOptions.Required = true;
+
+                        if (typeof(IBinding).IsAssignableFrom(propertyType))
+                            dotvvmProperty.MarkupOptions.AllowHardCodedValue = false;
+                        else if (!typeof(ValueOrBinding).IsAssignableFrom(propertyType.UnwrapNullableType()))
+                            dotvvmProperty.MarkupOptions.AllowBinding = false;
```

Comment:
This is problematic, if you add `MarkupOptions` to change the name or mapping mode, you'll make the property optional/bindable even when it must not be. Or you add it to make the property required, and suddenly it's not mapped as InnerElement

Reply:
So the `Required`, `AllowHardCodedValue`, and `AllowBinding` properties shall not be overridden by the attribute, while the other properties can be. I wonder whether it is intuitive, but I guess it's better than forcing the user to specify all these properties whenever they decide to use the attribute.

Reply:
No it's not intuitive, but we don't have a way to tell if the user actually set the property, or only used the attribute for other reason... It's not great that when you apply `[MarkupOptions(Required=true)]`, it will disable MappingMode.InnerElement :facepalm: